packages/apputils/src/printing.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * Any object is printable if it implements the `IPrintable` interface.
+ *
+ * If you have a custom widget and would like to make it printable, then you
+ * should fulfil that interface on that widget by adding a `[printSymbol]` method.
+ *
+ * One way of printing is to use the `printWidget` function, which creates a hidden iframe
+ * and copies the DOM nodes from your widget to that iframe and printing just that iframe.
+ *
+ * To use this method in your widget, set the `[printSymbol]` method to `printWidget`.
+ * You can also provide custom args to that function, including CSS text to apply
+ * when printing.
+ *
+ * See https://github.com/joseluisq/printd#print
+ *
+ * Another way to print is to use the `printURL` function, which takes a URL and prints that page.
+ */
+
+import { Printd, PrintdCallback } from 'printd';
+import { Widget } from '@phosphor/widgets';
+
+/**
+ * Symbol to use for a method that prints out out the object.
+ */
+export const printSymbol = Symbol();
+
+/**
+ * Widgets who provide a custom way of printing themselves
+ * should implement this interface.
+ */
+export interface IPrintable {
+  [printSymbol]: () => void;
+}
+
+/**
+ * Returns whether an object implements a print method.
+ */
+export function isPrintable(a: any): a is IPrintable {
+  try {
+    return printSymbol in a;
+  } catch {
+    // `in` raises a type error on non objects.
+    return false;
+  }
+}
+
+/**
+ * Calls the print symbol on an object to print it.
+ */
+export function print(a: IPrintable) {
+  a[printSymbol]();
+}
+
+/**
+ * Sets the print method on the parent to that of the child, if it
+ * exists on the child.
+ */
+export function deferPrinting(parent: IPrintable, child: object) {
+  if (isPrintable(child)) {
+    parent[printSymbol] = child[printSymbol].bind(child);
```

Comment:
This takes a snapshot of the child's current print method. Do you think it might be better to resolve the child print at runtime, with something like `parent[printSymbol] = () => child[printSymbol].call(args)`? It means the isPrintable check isn't as useful, but it does let the child swap out print methods.

Also, there's nothing here that forces a parent/child relationship. Perhaps just `delegatePrinting(self, other)` or something, with no mention of parent/child?

Reply:
> It means the isPrintable check isn't as useful, but it does let the child swap out print methods.

I need the `isPrintable` to work for it to know whether to grey it out in the menu. 

> Also, there's nothing here that forces a parent/child relationship. Perhaps just delegatePrinting(self, other) or something, with no mention of parent/child?

 Sounds good.

Reply:
I fixed this by having you return either a function or null from your print function. This allows you to implement the printable interface without being able to be printed. So this allows parents to more easily defer to their children's print functions, without having to do this kind of thing.